data-prepper-plugins/saas-source-plugins/jira-source/src/main/java/org/opensearch/dataprepper/plugins/source/jira/JiraService.java

```diff
@@ -0,0 +1,361 @@
+                       JiraSourceConfig jiraSourceConfig,
+                       JiraAuthConfig authConfig) {
+        this.restTemplate = restTemplate;
+        this.jiraSourceConfig = jiraSourceConfig;
+
+        issuesRequestedCounter = jiraPluginMetrics.counter(ISSUES_REQUESTED);
+        ticketFetchLatencyTimer = jiraPluginMetrics.timer(TICKET_FETCH_LATENCY_TIMER);
+        searchResultsFoundCounter = jiraPluginMetrics.counter(SEARCH_RESULTS_FOUND);
+        searchCallLatencyTimer = jiraPluginMetrics.timer(SEARCH_CALL_LATENCY_TIMER);
+        this.authConfig = authConfig;
+
+    }
+
+    /**
+     * Get jira entities.
+     *
+     * @param configuration the configuration.
+     * @param timestamp     timestamp.
+     * @param itemInfoQueue Item info queue.
+     */
+    public void getJiraEntities(JiraSourceConfig configuration, Instant timestamp,
+                                Queue<ItemInfo> itemInfoQueue) {
+        log.info("Started to fetch entities");
+        jiraProjectCache.clear();
+        searchForNewTicketsAndAddToQueue(configuration, timestamp, itemInfoQueue);
+        log.trace("Creating item information and adding in queue");
+        jiraProjectCache.keySet().forEach(key -> {
+            Map<String, Object> metadata = new HashMap<>();
+            metadata.put(CONTENT_TYPE, JiraContentType.PROJECT.getType());
+            ItemInfo itemInfo = createItemInfo(_PROJECT + key, metadata);
+            itemInfoQueue.add(itemInfo);
+        });
+
+    }
+
+    /**
+     * Method for building Issue Item Info.
+     *
+     * @param configuration Input Parameter
+     * @param timestamp     Input Parameter
+     */
+    private void searchForNewTicketsAndAddToQueue(JiraSourceConfig configuration, Instant timestamp,
+                                                  Queue<ItemInfo> itemInfoQueue) {
+        log.trace("Looking for Add/Modified tickets with a Search API call");
+        StringBuilder jql = createIssueFilterCriteria(configuration, timestamp);
+        int total;
+        int startAt = 0;
+        do {
+            SearchResults searchIssues = getAllIssues(jql, startAt, configuration);
+            List<IssueBean> issueList = new ArrayList<>(searchIssues.getIssues());
+            total = searchIssues.getTotal();
+            startAt += searchIssues.getIssues().size();
+            addItemsToQueue(issueList, itemInfoQueue);
+        } while (startAt < total);
+        searchResultsFoundCounter.increment(total);
+        log.info("Number of tickets found in search api call: {}", total);
+
+    }
+
+    /**
+     * Add items to queue.
+     *
+     * @param issueList     Issue list.
+     * @param itemInfoQueue Item info queue.
+     */
+    private void addItemsToQueue(List<IssueBean> issueList, Queue<ItemInfo> itemInfoQueue) {
+        issueList.forEach(issue -> {
+            Map<String, Object> issueMetadata = new HashMap<>();
+            if (Objects.nonNull(((Map) issue.getFields().get(PROJECT)).get(KEY))) {
```

Comment:
Having a model class will help here. This code is very complicated and difficult to work through.

Reply:
moved logic into JiraItemINfo Class